# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](https://github.com/spqdot/lab-neural-networks/blob/master/your-code/tttboard.jpg?raw=1)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

### Read tic-tac-toe.csv into a dataframe.

In [148]:
# your code here
import pandas as pd
import numpy as np

df =pd.read_csv('tic-tac-toe.csv')
df.head()

,TL,TM,TR,ML,MM,MR,BL,BM,BR,class
0,x,x,x,x,o,o,x,o,o,True
1,x,x,x,x,o,o,o,x,o,True
2,x,x,x,x,o,o,o,o,x,True
3,x,x,x,x,o,o,o,b,b,True
4,x,x,x,x,o,o,b,o,b,True


### Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.

In [149]:
df.shape

(958, 10)

In [150]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   TL      958 non-null    object
 1   TM      958 non-null    object
 2   TR      958 non-null    object
 3   ML      958 non-null    object
 4   MM      958 non-null    object
 5   MR      958 non-null    object
 6   BL      958 non-null    object
 7   BM      958 non-null    object
 8   BR      958 non-null    object
 9   class   958 non-null    bool  
dtypes: bool(1), object(9)
memory usage: 68.4+ KB


### Convert the categorical values to numeric in all columns.

In [151]:
df =df.replace('x',1)
df =df.replace('o',-1)
df =df.replace('b',0)
df.head()

/tmp/ipykernel_3321/1940231681.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df =df.replace('b',0)


,TL,TM,TR,ML,MM,MR,BL,BM,BR,class
0,1,1,1,1,-1,-1,1,-1,-1,True
1,1,1,1,1,-1,-1,-1,1,-1,True
2,1,1,1,1,-1,-1,-1,-1,1,True
3,1,1,1,1,-1,-1,-1,0,0,True
4,1,1,1,1,-1,-1,0,-1,0,True


In [152]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   TL      958 non-null    int64
 1   TM      958 non-null    int64
 2   TR      958 non-null    int64
 3   ML      958 non-null    int64
 4   MM      958 non-null    int64
 5   MR      958 non-null    int64
 6   BL      958 non-null    int64
 7   BM      958 non-null    int64
 8   BR      958 non-null    int64
 9   class   958 non-null    bool 
dtypes: bool(1), int64(9)
memory usage: 68.4 KB


In [153]:
df.isnull().sum()

,0
TL,0
TM,0
TR,0
ML,0
MM,0
MR,0
BL,0
BM,0
BR,0
class,0


### Separate the inputs and output

In [154]:
x= df.drop('class',axis=1)
y= df['class']

In [155]:
x.head()

,TL,TM,TR,ML,MM,MR,BL,BM,BR
0,1,1,1,1,-1,-1,1,-1,-1
1,1,1,1,1,-1,-1,-1,1,-1
2,1,1,1,1,-1,-1,-1,-1,1
3,1,1,1,1,-1,-1,-1,0,0
4,1,1,1,1,-1,-1,0,-1,0


In [156]:
y.head()

,class
0,True
1,True
2,True
3,True
4,True


### Normalize the input data.

In [157]:
### Normalize the input data
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
x_scaled = scaler.fit_transform(x)
x_scaled

array([[1., 1., 1., ..., 1., 0., 0.],
       [1., 1., 1., ..., 0., 1., 0.],
       [1., 1., 1., ..., 0., 0., 1.],
       ...,
       [0., 1., 0., ..., 1., 0., 1.],
       [0., 1., 0., ..., 1., 0., 1.],
       [0., 0., 1., ..., 0., 1., 1.]])

###

## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

### Split the training and test data.

In [158]:
# your code here
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.2, random_state=42)

### Create a Sequential model.

### Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.

In [159]:
num_classes = len(set(y))
print(num_classes)

2


In [160]:
### Create a Sequential model
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense

model = Sequential([
    Input(shape=(x_train.shape[1],)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(num_classes, activation='softmax')
])






### Compile the model using adam as the optimizer and sparse_categorical_crossentropy as the loss function. For metrics, use accuracy for now.

In [161]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

### Fit the training data.

In [162]:
history = model.fit(
    x_train,
    y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.2
)

Epoch 1/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.6275 - loss: 0.6487 - val_accuracy: 0.6818 - val_loss: 0.5992
Epoch 2/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6601 - loss: 0.5901 - val_accuracy: 0.7532 - val_loss: 0.5626
Epoch 3/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7288 - loss: 0.5439 - val_accuracy: 0.7792 - val_loss: 0.5295
Epoch 4/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7614 - loss: 0.5008 - val_accuracy: 0.7727 - val_loss: 0.5149
Epoch 5/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7712 - loss: 0.4603 - val_accuracy: 0.7987 - val_loss: 0.4660
Epoch 6/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8186 - loss: 0.4105 - val_accuracy: 0.8117 - val_loss: 0.4397
Epoch 7/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8546 - loss: 0.3724 - val_accuracy: 0.8247 - val_loss: 0.4017
Epoch 8/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8873 - loss: 0.3169 - val_accuracy: 0.8766 - val_loss

### Evaluate your neural network model with the test data.

In [163]:
loss, accuracy = model.evaluate(x_test, y_test)

print("Test Loss:", loss)
print("Test Accuracy:", accuracy)

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.9792 - loss: 0.0846 
Test Loss: 0.08455891162157059
Test Accuracy: 0.9791666865348816


### Save your model as tic-tac-toe.model.

In [164]:
model.save('tic-tac-toe.keras')

In [165]:
from tensorflow.keras.models import load_model

model = load_model('tic-tac-toe.keras')

## Step 3: Make Predictions
Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [166]:
### Select a few random rows from the test set
import numpy as np

random_idx = np.random.choice(len(x_test), size=5, replace=False)

X_sample = x_test[random_idx]
y_actual = y_test.iloc[random_idx]



In [167]:
pred_probs = model.predict(X_sample)
pred_classes = np.argmax(pred_probs, axis=1)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 82ms/step


In [168]:
print("Actual:", y_actual.values)
print("Predicted:", pred_classes)
print("Correct:", y_actual.values == pred_classes)

Actual: [False  True  True  True  True]
Predicted: [0 1 1 1 1]
Correct: [ True  True  True  True  True]


## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

### Add more layers

In [169]:
# your code here
model_1 = Sequential([
    Input(shape=(x_train.shape[1],)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(num_classes, activation='softmax')
])




In [170]:
model_1.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [171]:
history_1 = model_1.fit(
    x_train,
    y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.2
)

Epoch 1/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.6209 - loss: 0.6395 - val_accuracy: 0.7273 - val_loss: 0.5732
Epoch 2/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7271 - loss: 0.5467 - val_accuracy: 0.7143 - val_loss: 0.5501
Epoch 3/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7386 - loss: 0.5224 - val_accuracy: 0.6104 - val_loss: 0.6197
Epoch 4/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.7614 - loss: 0.4993 - val_accuracy: 0.7857 - val_loss: 0.5049
Epoch 5/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.7990 - loss: 0.4366 - val_accuracy: 0.7727 - val_loss: 0.4786
Epoch 6/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8268 - loss: 0.3899 - val_accuracy: 0.8636 - val_loss: 0.4054
Epoch 7/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.9036 - loss: 0.2774 - val_accuracy: 0.8766 - val_loss: 0.2689
Epoch 8/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.9657 - loss: 0.1624 - val_accuracy: 0.9675 - val

In [172]:
loss_1, accuracy_1 = model_1.evaluate(x_test, y_test)

print("Test Loss:", loss_1)
print("Test Accuracy:", accuracy_1)

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9948 - loss: 0.0314 
Test Loss: 0.03142702206969261
Test Accuracy: 0.9947916865348816


### Change the Learning Rate

In [173]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense

model = Sequential([
    Input(shape=(x_train.shape[1],)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(num_classes, activation='softmax')
])

In [174]:
import tensorflow as tf

optimizer = tf.keras.optimizers.Adam(
    learning_rate=0.001
)

In [175]:
model.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [176]:
history_2 = model.fit(
    x_train,
    y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.2
)

Epoch 1/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 16ms/step - accuracy: 0.6454 - loss: 0.6157 - val_accuracy: 0.6818 - val_loss: 0.5662
Epoch 2/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7239 - loss: 0.5536 - val_accuracy: 0.7662 - val_loss: 0.5381
Epoch 3/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7533 - loss: 0.5139 - val_accuracy: 0.7597 - val_loss: 0.5187
Epoch 4/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7729 - loss: 0.4780 - val_accuracy: 0.7662 - val_loss: 0.4867
Epoch 5/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7778 - loss: 0.4419 - val_accuracy: 0.7857 - val_loss: 0.4650
Epoch 6/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8023 - loss: 0.4205 - val_accuracy: 0.8182 - val_loss: 0.4265
Epoch 7/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8480 - loss: 0.3740 - val_accuracy: 0.8506 - val_loss: 0.3961
Epoch 8/20
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8725 - loss: 0.3337 - val_accuracy: 0.8896 - val_loss

In [177]:
loss_2, accuracy_2 = model.evaluate(x_test, y_test)

print("Test Loss:", loss_2)
print("Test Accuracy:", accuracy_2)

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.9792 - loss: 0.0821 
Test Loss: 0.08208069205284119
Test Accuracy: 0.9791666865348816


### Increase the Number of Epochs



In [178]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense

model = Sequential([
    Input(shape=(x_train.shape[1],)),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(32, activation='relu'),
    Dense(num_classes, activation='softmax')
])

In [179]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

In [180]:
history_3 = model.fit(
    x_train,
    y_train,
    epochs=50,
    batch_size=32,
    validation_split=0.2
)

Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.6471 - loss: 0.6127 - val_accuracy: 0.7143 - val_loss: 0.5809
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7141 - loss: 0.5609 - val_accuracy: 0.7792 - val_loss: 0.5497
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7647 - loss: 0.5135 - val_accuracy: 0.7597 - val_loss: 0.5277
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7663 - loss: 0.4786 - val_accuracy: 0.7662 - val_loss: 0.5098
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7876 - loss: 0.4457 - val_accuracy: 0.7922 - val_loss: 0.4682
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8252 - loss: 0.4009 - val_accuracy: 0.8117 - val_loss: 0.4365
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8595 - loss: 0.3530 - val_accuracy: 0.7987 - val_loss: 0.4223
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.9003 - loss: 0.3085 - val_accuracy: 0.8636 - val_loss

In [181]:
loss_3, accuracy_3 = model.evaluate(x_test, y_test)

print("Loss:", loss_3)
print("Accuracy:", accuracy_3)

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.9896 - loss: 0.0470 
Loss: 0.046996671706438065
Accuracy: 0.9895833134651184


**Which approach(es) did you find helpful to improve your model performance?**

### Adding more layers and increasing the number of training epochs improved the model's performance. The best results were achieved by adding more layers, which reduced the test loss to 0.0314 and increased the test accuracy to 99.48%. Adjusting the learning rate did not significantly improve performance, suggesting that the default learning rate was already appropriate for this dataset.